Source/WebCore/style/AnchorPositionEvaluator.cpp

```diff
@@ -510,37 +502,87 @@ static bool isAcceptableAnchorElement(Ref<const Element> anchorElement, Ref<cons
-    for (auto anchor : makeReversedRange(anchors)) {
-        ASSERT(anchor->renderer());
+    for (auto& anchor : makeReversedRange(anchors)) {
         if (isAcceptableAnchorElement(anchor.get(), anchorPositionedElement))
-            return anchor.get();
+            return *anchor->element();
     }
 
     return { };
 }
 
-void AnchorPositionEvaluator::findAnchorsForAnchorPositionedElement(Ref<const Element> anchorPositionedElement)
+static AnchorsForAnchorName collectAnchorsForAnchorName(const Document& document)
+{
+    if (!document.renderView())
+        return { };
+
+    AnchorsForAnchorName anchorsForAnchorName;
+
+    auto& anchors = document.renderView()->anchors();
+    for (auto& anchorRenderer : anchors) {
+        for (auto& name : anchorRenderer.style().anchorNames()) {
+            anchorsForAnchorName.ensure(name, [&] {
+                return AnchorsForAnchorName::MappedType { };
+            }).iterator->value.append(anchorRenderer);
+        }
+    }
+
+    // Sort them in tree order.
+    for (auto& anchors : anchorsForAnchorName.values()) {
+        std::sort(anchors.begin(), anchors.end(), [](auto& a, auto& b) {
+            // FIXME: Figure out anonymous pseudo-elements.
+            if (!a->element() || !b->element())
+                return !!b->element();
+            return is_lt(treeOrder<ComposedTree>(*a->element(), *b->element()));
+        });
+    }
+
+    return anchorsForAnchorName;
+}
+
+AnchorElements AnchorPositionEvaluator::findAnchorsForAnchorPositionedElement(const Element& anchorPositionedElement, const HashSet<AtomString>& anchorNames, const AnchorsForAnchorName& anchorsForAnchorName)
 {
-    auto* anchorPositionedState = anchorPositionedElement->document().styleScope().anchorPositionedStates().get(anchorPositionedElement);
-    ASSERT(anchorPositionedState && anchorPositionedState->stage == AnchorPositionResolutionStage::FinishedCollectingAnchorNames);
+    AnchorElements anchorElements;
 
-    for (auto& anchorName : anchorPositionedState->anchorNames) {
-        auto anchor = findLastAcceptableAnchorWithName(anchorName, anchorPositionedElement);
-        if (anchor.has_value())
-            anchorPositionedState->anchorElements.add(anchorName, anchor->get());
+    for (auto& anchorName : anchorNames) {
+        auto anchor = findLastAcceptableAnchorWithName(anchorName, anchorPositionedElement, anchorsForAnchorName);
+        if (anchor)
+            anchorElements.add(anchorName, anchor->get());
     }
 
-    anchorPositionedState->stage = AnchorPositionResolutionStage::FoundAnchors;
+    return anchorElements;
+}
+
+void AnchorPositionEvaluator::updateAnchorPositioningStatesAfterLayout(const Document& document)
+{
+    std::optional<AnchorsForAnchorName> anchorsForAnchorName;
+    for (auto elementAndState : document.styleScope().anchorPositionedStates()) {
+        if (!anchorsForAnchorName)
+            anchorsForAnchorName = collectAnchorsForAnchorName(document);
```

Comment:
We can just collect anchors before the loop right? Or this is to avoid doing any work if there's no anchor positioned elements?

Reply:
Yeah, this avoid collection in the (common) no-anchors case

Reply:
Changed it to simply test if the map is empty first.